File: src/services/book/bookIdentity.ts
const CONTROL_CHAR_RANGE = '\\u0000-\\u001f'
const INVALID_FILENAME_CHARS = new RegExp(`[<>:"/\\\\|?*${CONTROL_CHAR_RANGE}]`, 'g')
const MULTIPLE_SPACES = /\s+/g

export const normalizeBookIdentityPart = (value?: string, fallback = 'unknown'): string => {
  const sanitized = (value || '')
    .replace(INVALID_FILENAME_CHARS, '_')
    .replace(MULTIPLE_SPACES, ' ')
    .trim()

  return sanitized || fallback
}

export const buildBookTitle = (title?: string): string => {
  return normalizeBookIdentityPart(title, 'untitled')
}

export const buildBookName = (title?: string, author?: string): string => {
  const safeTitle = normalizeBookIdentityPart(title, 'untitled')
  const safeAuthor = normalizeBookIdentityPart(author, 'unknown')

  return `${safeTitle}_${safeAuthor}`
}

export const toBookConfigFilename = (bookKey: string): string => `${bookKey}.json`

export const getBookKeyFromConfigFilename = (filename: string): string => {
  return filename.replace(/\.json$/i, '')
}

export const hashBookKey = async (bookKey: string): Promise<string> => {
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(bookKey))

  return Array.from(new Uint8Array(digest))
    .map((byte) => byte.toString(16).padStart(2, '0'))
    .join('')
}

File: src-tauri/src/service/book_identity.rs
use sha2::{Digest, Sha256};

const INVALID_FILENAME_CHARS: [char; 9] = ['<', '>', ':', '"', '/', '\\', '|', '?', '*'];

fn normalize_book_identity_part(value: Option<&str>, fallback: &str) -> String {
    let replaced: String = value
        .unwrap_or_default()
        .chars()
        .map(|ch| {
            if ch.is_control() || INVALID_FILENAME_CHARS.contains(&ch) {
                '_'
            } else {
                ch
            }
        })
        .collect();

    let normalized = replaced.split_whitespace().collect::<Vec<_>>().join(" ");
    if normalized.is_empty() {
        fallback.to_string()
    } else {
        normalized
    }
}

pub fn build_book_title(title: Option<&str>) -> String {
    normalize_book_identity_part(title, "untitled")
}

pub fn build_book_key(title: Option<&str>, author: Option<&str>) -> String {
    let title = normalize_book_identity_part(title, "untitled");
    let author = normalize_book_identity_part(author, "unknown");

    format!("{}_{}", title, author)
}

/// 缓存目录名使用 SHA-256 前 8 字节（16 位十六进制）缩短后的哈希，
/// 在保证书籍库场景下碰撞概率可忽略的同时避免目录名过长。
const HASH_BYTE_LEN: usize = 8;

pub fn hash_book_key(book_key: &str) -> String {
    let digest = Sha256::digest(book_key.as_bytes());
    digest[..HASH_BYTE_LEN]
        .iter()
        .map(|byte| format!("{:02x}", byte))
        .collect()
}
